File: AlgorithmDataStructure/LeetCode/8-StringToInt/atoi.c

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)

int myAtoi(char * str){
    long length = strlen(str);
    if (length == 0) { return 0; }
    
    int sign = 1, flag = 0;
    long result = 0;
    for (int i = 0; i < length; i++) {
        char item = str[i];
        
        if (flag == 0) {
            if (item == '-') {
                sign = -1;
                flag = 1;
            } else if (item == '+') {
                flag = 1;
            } else if (item >= '0' && item <= '9') {
                result = item - 48;
                flag = 1;
            } else if (item == ' ') {
                continue;
            } else {
                return 0;
            }
        } else if (flag == 1) {
            if (item >= '0' && item <= '9') {
                result = result * 10 + item - 48;
                if (sign == 1 && result >= INT_MAX) {
                    return INT_MAX;
                } else if (sign == -1 && result - 1 >= INT_MAX) {
                    return INT_MIN;
                }
            } else {
                return (int)result * sign;
            }
        }
    }
    return (int)result * sign;
}

int main(int argc, const char * argv[]) {
    
    printf("==%d==\n", myAtoi("-2147483649"));
    //"  +42 ab 345"  42
    
    //"-2147483647"  -2147483647
    //"-2147483648"  -2147483648
    //"-2147483649"  -2147483648
    //"2147483648"  2147483647
    
    return 0;
}

File: AlgorithmDataStructure/LeetCode/8-StringToInt/atoi.swift

class Solution {
    
    let INT_MAX = 2147483647
    let INT_MIN = -2147483648
    
    
    //自动机
    //每个时刻有一个状态 s，每次从序列中输入一个字符 c，并根据字符 c 转移到下一个状态 s'。
    //这样只需要建立一个覆盖所有情况的从 s 与 c 映射到 s' 的表格即可解决题目中的问题。
    
    private var state = "start"
    private let map = ["start":["start", "signed", "in_number", "end"],
    "signed":["end", "end", "in_number", "end"],
    "in_number":["end", "end", "in_number", "end"],
    "end":["end", "end", "end", "end"]]
    
    var ans = 0
    var sign = 1
    
    func get_col(_ c: Character) -> Int {
        if c == " " { return 0; }
        if c == "+" || c == "-" { return 1 }
        let i = Int("\(c)")
        if i != nil {
            if  i! >= 0 && i! <= 9 { return 2 }
        }
        return 3
    }
    
    func get(_ c: Character) {
        state = map[state]![get_col(c)]
        if state == "in_number" {
            ans = ans * 10 + Int("\(c)")!
            ans = sign == 1 ? min(ans, INT_MAX) : min(ans,-1 * INT_MIN)
        } else if state == "signed" {
            sign = c == "+" ? 1 : -1
        }
    }
    
    
    func myAtoi(_ str: String) -> Int {
        for c in str {
            get(c)
        }
        return sign * ans
    }
}
let s = Solution()
print(s.myAtoi("-2147483648"))
